### PR TITLE
feat: add end to end identity bindings [CL-21]

### PR DESCRIPTION
**Pull request by beltram**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

Add end to end identity bindings for Android. WASM support is partial and to be continued. Swift bindings are not done.

----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: crypto-ffi/src/wasm.rs [hunk 1]
```diff
@@ -1584,10 +1583,187 @@ impl CoreCrypto {
...
+    }
+
+    /*pub fn new_order_response(&self, order: Uint8Array) -> WasmCryptoResult<Uint8Array> {
+        let order: Vec<u8> = self.0.new_order_response(order.to_vec())?.into();
+        WasmCryptoResult::Ok(order.as_slice().into())
+    }*/
+}
+
+#[wasm_bindgen]
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
```
Comment (OtaK) [inline crypto-ffi/src/wasm.rs:1658]: Unsure but you might need to add something like `#[serde(rename_all = "camelCase")]` for the JSON conversion to happen correctly between the JS naming schemes and the rust-native `snake_case`

File: crypto/src/e2e_identity/types.rs [hunk 2]
```diff
@@ -0,0 +1,212 @@
+//! We only expose byte arrays through the FFI so we do all the conversions here
+
+use super::error::{E2eIdentityError, E2eIdentityResult};
+
+/// TODO
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+pub struct E2eiAcmeDirectory {
```
Comment (OtaK) [inline crypto/src/e2e_identity/types.rs:7]: same here as with the TS/WASM conversions, you might need a `#[serde(rename_all = "camelCase")]` directive

File: crypto/src/e2e_identity/types.rs [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+                },
+            )?,
+            new_order: serde_json::to_value(new_order.delegate)?,
+        })
+    }
+}
+
+/// TODO
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+pub struct E2eiNewAcmeAuthz {
```
Comment (OtaK) [inline crypto/src/e2e_identity/types.rs:97]: Same possible JSON serialization issues with the casing here

**Review by OtaK — CHANGES_REQUESTED** [review]

File: crypto-ffi/src/generic.rs [hunk 4]
```diff
@@ -979,3 +991,259 @@ impl CoreCrypto<'_> {
...
+    }
+
+    pub fn new_authz_response(&self, authz: Json) -> E2eIdentityResult<NewAuthz> {
+        Ok(self.0.new_authz_response(authz)?.into())
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    pub fn create_dpop_token(
+        &self,
+        access_token_url: String,
```
[the remark this answers is not in the document]
Comment (beltram) [reply crypto-ffi/src/generic.rs:1045]: Hmmm that would require maintaining yet another struct across the FFI which might be cumbersome to maintain

**Pull request merged**